### DISCUSSION TITLE
Variation in firing of L2/3 Pyramidal cells between trials

### DISCUSSION THREAD

**Discussion by mkhalil8** [jonescompneurolab/hnn-core#632] [neuro]
Hi @jasmainak , @ntolley  and everyone,

what explains the big variability in L2/3 pyramidal firing with trials? and why their firing that is usually around 120msecs (supposedly caused by the Distal input) is delayed from the mean of most of the distal input?
I understand that the Vecstim objects connectivity to the different populations of cells is distributed and that the main change in between trials is the seeds (incremented by 1 as I remember) which changes the timing of the spiking input.
These are 25 trials using Kohl's auditory ERP parameters.
 
![Screen Shot 2023-03-31 at 1 13 08 PM](https://user-images.githubusercontent.com/14954705/229186352-f8667b8c-7acc-495f-8139-83769ab2336b.png)

**Comment by rythorpe** [comment]
The variability you're observing is caused by local network interactions. The local network connectivity to and from L2/3 pyramidal cells (specifically the L2/3 basket -> L2/3 pyramidal connections) makes their firing particularly sensitive to small variations in exogenous drive timing. Notice the following sequence of events:

1.  an initial barrage of L2/3 pyramidal spikes (dark green I'm assuming) occurs at ~50 ms robustly across trials
2. L2/3 pyramidal firing is quickly squashed by inhibition from L2/3 basket cell firing (light green I'm assuming)
3. L2/3 basket cells fire a little here and there from 50-100 ms, and are easily excited by exogenous drive
4. L2/3 pyramidal cells are mostly silenced from 50-100 ms, unless they are transiently released from L2/3 basket inhibition or until the second proximal drive starts driving L2/3 with excitation

Altogether, L2/3 pyramidal spiking rests right near the threshold of local network E/I balance. The tight coupling between L2/3 pyramidal and L2/3 basket cells also allows for pyramidal interneuron gamma (PING) activity to emerge transiently. All that said, I encourage you to play around with the local network connectivity parameters in order to test my assertions above (e.g., try increasing the late proximal drive timing and L2/3 pyramidal weight to test the hypothesis that the ~120 ms barrage of L2/3 pyramidal spikes shifts right and with greater temporal consistency/precision).

Note: updated on 4/3/23

**Comment by rythorpe** [comment]
I'm going to convert this to a discussion, but feel free to open another issue if you end up finding a bug.

**Comment by mkhalil8** [comment]
Thank you very much @rythorpe, this is really helpful and got me to think in different direction. So as I imagine it now, on a single L2/3 pyramidal level, after firing initially and then inhibited, it continues to receive inputs that are both excitatory and inhibitory but the inhibitory tone is higher so it remains mainly without firing till it reaches a point of getting more excitatory inputs from the L5pyramidal cells which in general are more capable of firing due to their connectivity pattern with their basket cells. The sensitivity to timing of exogenous inputs you mentioned is due to the state of the network when the input arrives?

**Reply by rythorpe** [reply]
Yes, exactly. Note also that L2/3 pyramidal have attenuating spiking properties in response to a tonic current injection while L5 pyramidal have bursting spiking properties. This, in addition to the distinct connectivity patterns for each respective cell type, contributes to the differential properties you observe between L2/3 pyramidal and L5 pyramidal cells.

**Reply by mkhalil8** [reply]
Again, Thank you @rythorpe!! I am reading now about bursting behavior and PING (please don't hesitate to share any resources that you feel necessary to understand the subject, especially relevant to modeling them).
Based on what I understood so far the bursting behavior in pre-synaptic cell can affect the post synaptic cell differently compared to a single spike(It accumulates post synaptic potential and the frequency of the bursting interacts with subthreshold oscillations in post synaptic cells if present), it also can facilitate or depress the synapse, also based on inter-burst frequency.
I can see the accumulation in post synaptic potential is modeled in HNN-Core especially when I plot the voltage of the Pyramidal cells sections, however, I don't believe that the synapse facilitation or depression between L5 and L2/3 pyramidal through change in weight is modeled here? right?

**Reply by rythorpe** [reply]
If you're interested in learning more about PING, two good papers that come to mind are [Vierling-Claassen 2010](https://doi.org/10.3389/fnhum.2010.00198) and [Tiesinga & Sejnowski 2009](https://doi.org/10.1016/j.neuron.2009.09.009).

Re: synaptic facilitation/depression, you're correct that this is something we don't explicitly include in the model (i.e., all weights are fixed).

**Comment by mkhalil8** [comment]
@rythorpe what are the synaptic connections from L5pyr to L2/3pyr? I can't find any in the param files?

**Reply by rythorpe** [reply]
Hi @mkhalil8, you can use the [`Network.connectivity`](https://jonescompneurolab.github.io/hnn-core/stable/generated/hnn_core.Network.html#hnn_core.Network) attribute to access the properties of each connection type. You'll want to first select the correct connection index using [`pick_connection`](https://jonescompneurolab.github.io/hnn-core/stable/generated/hnn_core.pick_connection.html#hnn_core.pick_connection) and then use that index to retrieve the desired connection:
 ```python
conn_idxs = pick_connnection(net=net, src_gids='L5_pyramidal', target_gids='L2_basket')   # returns a list of indices
print(net.connectivity[conn_idxs[0])  # view the properties of the first conn_idx
```
See our ["Modifying local connectivity"](https://jonescompneurolab.github.io/hnn-core/stable/auto_examples/howto/plot_connectivity.html#sphx-glr-auto-examples-howto-plot-connectivity-py) tutorial for more details.

**Reply by mkhalil8** [reply]
I don't think there are direct connections from L5pyr to L2pyr, I wonder how does the excitatory activity transmits from L5pyr to L2pyr?

**Reply by rythorpe** [reply]
Oh, I see what you mean. Good catch - I'm not sure why I thought there was ever a L5pyr -> L2pyr connection because I don't think any of our models have these :man_facepalming:. The ~125 ms excitation might be coming to L2pyr from the earliest of the late distal drive (proximal 2) events, since this drive is very spread out. If the L2pyr synaptic weight of this drive is strong enough, it will propagate among L2pyr, but then quickly spread to L2basket which then reciprocate with inhibition.

**Comment by mkhalil8** [comment]
Ok, I see. So to answer the question about the mechanism of the ~125msecs L2/3pyr firing I had two possibilities in mind from our discussion, first is that the same mechanism you described happens but with the distal drive accumulating excitatory inputs(instead of L5 inputs) till it overcome the inhibitory tone.

The second possibility is that the firing is mainly due to the proximal 2 drive (which has stronger connection to L2/3pyr than distal drive) especially in the intersection between d1 & p2.

I looked closely at the L2/3pyr firing in different trials with the histograms associated with it and looks like the firing does not happen till the P2 arrives, so it is mainly the second possibility and not the first one. 
I tested whether the firing is purely p2 or with the contribution of distal 1 by looking at four rasters which has the same seed but with increasing the AMPA weight of the distal drive to all cells by different values, and this has led to stronger 125ms firing.

So in conclusion, I think that the L2/3pyr firing at ~125msecs is due to initial sub-threshold excitation from the distal drive then the proximal 2 drive arrives.

**Comment by mkhalil8** [comment]
Hi Ryan @rythorpe , hope all is good :)
I have a question about the connectivity of the network, is there randomness in the connections each time you instantiate a network. Or all instances have the same connectivity in between cells and the main variance will be the drives

**Reply by rythorpe** [reply]
Hi @mkhalil8, if you instantiate the network using one of our template models (e.g., `jones_2009_model()`), the connectivity is all-to-all and hence not random. If you create your own model version with <1 connection probability between any two source/target connection types, you're given the option to specify the [`conn_seed`](https://github.com/jonescompneurolab/hnn-core/blob/400d1b75f3b5717576e4feb335941e3d7918d7fe/hnn_core/network.py#L1119) which will enforce a single set of pseudorandom connections that will be replicated every time you instantiate a new network. If you don't specify `conn_seed`, the default behavior is to randomize connectivity differently for each instantiation.
